stratum/hal/lib/common/p4_service.cc

```diff
@@ -384,18 +384,6 @@ ::grpc::Status P4Service::SetForwardingPipelineConfig(
                      node_id, "."));
   }
 
-#ifdef P4OVS_CHANGES
-  // FIXME: We do not support overwrite of an already configured pipeline
-  // for a single device. Remove this check when support for overwriting an
-  // already configured forwarding pipeline is available.
-  if (forwarding_pipeline_configs_ != nullptr &&
-      forwarding_pipeline_configs_->node_id_to_config_size() != 0) {
-    return ::grpc::Status(::grpc::StatusCode::FAILED_PRECONDITION,
-                          "Only a single forwarding pipeline can be pushed "
-                          "for any node so far.");
-  }
```

Comment:
Does tdi code(previously bfrt_node.cc) handle overwriting the pipeline config?

Reply:
Which layer of overwrite are we talking about?
At the p4_service.cc layer, it seems to write to a file in /etc/stratum/pipeline_config.pb.txt.
Beyond this, the warm init APIs are invoked to flash a new pipeline into the HW. bfrt_node is where this overwrite happens.

Reply:
> Please change P4OVS_CHANGES to DPDK_TARGET.
> 
> Otherwise, we're going to have to put the code back in again in a couple of weeks for DPDK.

This is new behavior different from the original Stratum if we add the DPDK specific check.
If DPDK does not support setForwardingPipeline, let the error come from within the DPDK SDK and propagated back.

Reply:
This is new code, added in P4-OVS before TDI was implemented.

Reply:
> Which layer of overwrite are we talking about? At the p4_service.cc layer, it seems to write to a file in /etc/stratum/pipeline_config.pb.txt. Beyond this, the warm init APIs are invoked to flash a new pipeline into the HW. bfrt_node is where this overwrite happens.

Talking about bfrt_node overwrite

Reply:
My point is the old P4-OVS changes shouldn't matter. We cannot have target specific changes outside of hal/lib/tdi.

Reply:
Agreed to the point that DPDK SDE should return NOT_SUPPORTED. I have never tested overwriting the config. My question was more about does Tofino target support this and is it well tested feature?

Reply:
Yes. Tofino supports setForwardingPipeline from day one.

Reply:
Right now, TDI is not capable of subsuming all target-specific aspects of Stratum. If this were true, there would be no need to have #ifdefs in tdi_sde_wrapper or tdi_id_mapper.

Your "let DPDK take care of it" comment assumes that DPDK has enough deep knowledge of Stratum to detect the overwrite condition itself. That doesn't strike me as realistic, any more than the belief that TDI hides all target differences.

For the interim, we need to use #ifdefs so we can meet our deliverables.

Changing the conditional from P4OVS_CHANGES to DPDK_TARGET will resolve your Tofino problem without breaking DPDK in the process. Please do not break DPDK.

Reply:
setForwardingPipeline is not a Stratum defined behavior. It is a P4Runtime API and generally considered a foundational API within P4Runtime.

The existing change is a hack made for DPDK and the correct behavior still is to address it from within the SDK. I don't think we are breaking DPDK but actually ensuring that DPDK conforms to the P4Runtime standard.